### PR TITLE
Error if rpcpassword contains hash in conf sections

### LINKED ISSUE #15075: Expected error not printed if rpcpassword contains hash character but is in a [section]
I'm writing a quick release note for #14494 (Error if # is used in rpcpassword in conf) and it doesn't work the way I expect with the optional new configuration file style.  For example, this configuration file aborts startup as expected:

```text
$ cat .bitcoin/bitcoin.conf ; bitcoind -testnet | tee /dev/null | head -n 5
rpcpassword=foo#bar
Error reading configuration file: parse error on line 1, using # in rpcpassword can be ambiguous and should be avoided
```

However, this does not:

```text
harding@redspot:~$ cat .bitcoin/bitcoin.conf ; bitcoind -testnet | tee /dev/null | head -n 5
[test]
rpcpassword=foo#bar
2019-01-01T20:24:25Z Bitcoin Core version v0.17.99.0-f5a70d1462 (release build)
2019-01-01T20:24:25Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2019-01-01T20:24:25Z Assuming ancestors of block 0000000000000037a8cd3e06cd5edbfe9dd1dbcc5dacab279376ef7cfc2b4c75 have valid signatures.
2019-01-01T20:24:25Z Setting nMinimumChainWork=00000000000000000000000000000000000000000000007dbe94253893cbd463
2019-01-01T20:24:25Z Using the 'sse4(1way),sse41(4way)' SHA256 implementation
```

The same problem with rpcpassword applies to the `[main]` section. I can confirm that options specified in sections are otherwise working correctly here, and that rpcpassword also otherwise works as expected---I only fail to get the expected error when putting it in a section.

I'm running commit f5a70d1462592a23bbad4aa150e6b2beaeec7c42 (master from 2019-01-01) on Debian Stable on x86_64.

CC: @MeshCollider (author of #14494)

### PR DESCRIPTION
**Pull request by meshcollider**
Fixes #15075

### CODE REVIEW

File: test/functional/feature_config_args.py [hunk 1]
```diff
@@ -30,8 +30,8 @@ def test_config_file_parser(self):
         self.nodes[0].assert_start_raises_init_error(expected_msg='Error reading configuration file: parse error on line 1: nono, if you intended to specify a negated option, use nono=1 instead')
 
         with open(inc_conf_file_path, 'w', encoding='utf-8') as conf:
-            conf.write('server=1\nrpcuser=someuser\nrpcpassword=some#pass')
-        self.nodes[0].assert_start_raises_init_error(expected_msg='Error reading configuration file: parse error on line 3, using # in rpcpassword can be ambiguous and should be avoided')
+            conf.write('server=1\nrpcuser=someuser\n[main]\nrpcpassword=some#pass')
```
Comment (laanwj) [inline test/functional/feature_config_args.py]: maybe keep one outside the section too

**Comment by laanwj** [comment]
utACK

**Comment by promag** [comment]
Tested ACK 170523f.

I prefer this version though:
```diff
diff --git a/src/util/system.cpp b/src/util/system.cpp
index ffe4d21ee..8da476289 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -863,12 +863,13 @@ static bool GetConfigOptions(std::istream& stream, std::string& error, std::vect
                 error = strprintf("parse error on line %i: %s, options in configuration file must be specified without leading -", linenr, str);
                 return false;
             } else if ((pos = str.find('=')) != std::string::npos) {
-                std::string name = prefix + TrimString(str.substr(0, pos), pattern);
-                std::string value = TrimString(str.substr(pos + 1), pattern);
-                if (used_hash && name.find("rpcpassword") != std::string::npos) {
+                std::string name = TrimString(str.substr(0, pos), pattern);
+                if (used_hash && name == "rpcpassword") {
                     error = strprintf("parse error on line %i, using # in rpcpassword can be ambiguous and should be avoided", linenr);
                     return false;
                 }
+                name = prefix + name;
+                std::string value = TrimString(str.substr(pos + 1), pattern);
                 options.emplace_back(name, value);
                 if ((pos = name.rfind('.')) != std::string::npos) {
                     sections.insert(name.substr(0, pos));
```

**Comment by laanwj** [comment]
@promag huh, yes, I hadn't noticed that the prefix was added there, I tend to agree that's better than a fuzzy query like `.find()`, I didn't consider it important enough to remark as I thought it involved increasing the amount/complexity of string parsing, but that doesn't seem too bad.

File: test/functional/feature_config_args.py (hunk 1 above)
Comment (laanwj) [reply test/functional/feature_config_args.py]: okay, it's clear we need to test all three, after the recent discussion it is very clear it's easy to miss one:
```
rpcpassword=   # bare
regtest.rpcpassword=   # explicit section

[regtest]
rpcpassword=   # implicit section
```

**Comment by laanwj** [comment]
re-utACK 8cff83124bcac936ecc6add6dca72b125a79a08f

**Comment by jnewbery** [comment]
tested ACK 8cff83124bcac936ecc6add6dca72b125a79a08f

**Comment by promag** [comment]
post merge ACK 8cff831.

**Comment by AkioNak** [comment]
Sorry for my late commet.
Is "rrrpcpassworddd=foo#bar" an error? 
 I think that the unknown configuration file options will be ignored. 
( as a compromise. - #13799)

**Pull request merged**